torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -1154,6 +1154,91 @@ std::vector<WorkerInfo> TensorPipeAgent::getWorkerInfos() const {
   return workerInfos;
 }
 
+void TensorPipeAgent::updateGroupMembership(
+    const WorkerInfo& workerInfo,
+    const std::vector<c10::Device> devices,
+    const std::unordered_map<std::string, DeviceMap> reverseDeviceMaps,
+    bool isJoin = true) {
+  std::cout << "in updateGroupMembership" << std::endl;
+  std::cout << workerInfo.name_ << std::endl;
+
+  std::string name = workerInfo.name_;
+  worker_id_t id = workerInfo.id_;
+  /*
+  TODO: update
+  workerIdToInfo_
+  workerNameToInfo_
+  workerNameToURL_
+  devices_
+  reverseDeviceMaps_
+  */
+  // Rank with workerInfo is joining the group, update internal mappings
+  if (isJoin) {
+    std::cout << "in isJoin" << std::endl;
+    workerIdToInfo_.emplace(id, workerInfo);
+    workerNameToInfo_.emplace(name, workerInfo);
+
+    // TODO: we should get nodeAddrStr in the joining process, then pass in as
+    // an argument rather than getting from store each time
+    auto nodeAddrData = nameToAddressStore_.get(name);
+    auto nodeAddrStr =
+        std::string((const char*)nodeAddrData.data(), nodeAddrData.size());
+    workerNameToURL_.insert({name, nodeAddrStr});
+
+    // TODO: update so that these mappings are locked
+    // if (!reverseDeviceMap.empty()) {
+    //   reverseDeviceMaps_.emplace(name, reverseDeviceMap);
+    //   for (const auto it : reverseDeviceMap) {
+    //     // push back each device if it does not exist
+    //     if (std::find(devices_.begin(), devices_.end(), it.first) ==
+    //         devices_.end()) {
+    //       devices_.push_back(it.first);
+    //     }
+    //   }
+    // }
+    reverseDeviceMaps_ = reverseDeviceMaps;
```

Comment:
curious, any reason that we need to update the entire `reverseDeviceMaps_ `, instead of just insert the info for the newly joined node? If this new node's info already exist in the `reverseDeviceMaps_`, it should return an error?

Reply:
Every time a new worker joins I am performing validation of all the device maps currently set by the group's workers. I am also constructing the reverseDeviceMap so I just passed it in here for simplicity. Now I updated the function to add the map only if the worker does not exist, in case of the new node info already existing, that error will be checked and thrown in `_validate_device_maps()`.